### DISCUSSION TITLE
My long time directus usage experiences for a medium size project

### DISCUSSION THREAD

**Discussion by emrahnazif** [directus/directus#14794] [Archive]
hello,

I am using directus for my project and I would like to share my experiences with details. It will be more like a medium articles but here it is...

First, directus is an extremely well architectured, flexible and I believe it is one of the best in the market. It can be used in many different projects. Thank you.

Despite many great features I would like to explain difficulties (that I had) using directus in a medium size project.
First I started to use directus in digitalocean, with a relatively decent server with s3 for files storage. That went unfortunately very unpleasant for me. I was uploading 1000 image files. When I try to download them in paralel (30-40 images per second), after a moment download processes was stoping with many errors. Probably because I was hitting digitalocean space rate limits, they block me for 10-15 mins, then the files starts to work again. digitalocean spaces limits are higher then what I do. I am not sure if somehow directus makes unnecessary calls to s3 that leads to be blocking. 

Then I move directus instance to an apple m1 with a nice ssd attached (I use local  storage). Above problem solved. But this breaks the scalability of my app. 

Now I have around 6 millions files stored in API. When I click the "files" from the App, it makes 2 request for files and it takes about 1 minutes to load the page. That is too long for a response. 

<img width="1360" alt="Screenshot 2022-08-02 at 12 41 56" src="https://user-images.githubusercontent.com/33692017/182356635-05710f2c-908a-463a-9485-760cd7176577.png">

Meanwhile, when I access files via API https://directus.api/files?access_token=access_token 
It is almost instant. So the problem is not the API but it is what information the APP is asking to the API. The more files I have, the more it's getting longer. Even right now app/files is not usable for me, because for every single click I had to wait 1 minute. 

Same continues for other collections as well. If you have too many items in a collection, waiting time is extremely high inside the app. I have a collection with 6m items in side, it tooks 20 secs to load the items inside the app. The problem is, it is taking longer and longer every day. 

File management:
If you have a few files, thats fine. Everything works great. If you have millions of files. It is a nightmare. One the reason that I abandon s3 was, deleting files was almost impossible. Because directus uses find&delete while deleting files which makes deleting a single file can take 60-100 secs. If you would like to delete 1000s files, good luck. That why I use many local storage adapters so files are distributed in many directories. It is a pain to manage, and it still take 15-20 secs to delete a file. So, right now, I don't delete files any more. The problem gets bigger and bigger for me. 

For deleting files or displaying a big collection, CPU usage hits its peak which makes every other operation suffer. 

if you are making updates in your collection/items, the activity table and revisions table grows very fast. Which is possible to disable Activity & Revision Tracking for user created collections but they are enabled by default for system collections and it is not possible to disable. To my knowlege, it is not possible to delete them from database easily. Today around 80% of my database size is thanks to Activity & Revision Tracking. I love it, it is something must have but it can be improved. 

6m items in a collection or 6m files are pretty small for a big project. 

Even though I only talked about the downsides for my use-case, directus is great, I flow flows and insights (although getting numbers takes too long), user management, community and support everything is great. 

Until now, I can say API part works pretty well, APP is not that great, it makes many things slower. File management should be better.

I will definitely continue using directus, it is great. however there is something makes me afraid about the scalability of the API/APP. What would it be like, if I have 100k monthly active users, how can I manage files when I reach 100m files, using the app if I have 50m items inside a collection etc. 

Again, 
thanks for all

**Comment by rijkvanzanten** [comment]
@emrahnazif Mind sharing the exact request that the app is doing on your end, and the exact number of files you have in that particular folder it's requesting? Curious to see if we can find the particular bottleneck in that app request, as it sounds like some of those filters the app is adding is causing the slowdown 🙂

**Reply by emrahnazif** [reply]
> @emrahnazif Mind sharing the exact request that the app is doing on your end, and the exact number of files you have in that particular folder it's requesting? Curious to see if we can find the particular bottleneck in that app request, as it sounds like some of those filters the app is adding is causing the slowdown 🙂

Thanks, created an issue for this. https://github.com/directus/directus/issues/14796

3 bottlenecks: sort, filter_count, total_count

**Reply by benhaynes** [reply]
Hey @emrahnazif — I just wanted to quickly add a _thank you_ for posting such a thoughtful and comprehensive Discussion here. It's constructive conversations like this that really help us push the platform forward. 👍
